### PR TITLE
Document directive sorting algorithm

### LINKED ISSUE #214: Document Caddyfile directive/matcher sorting behaviour
Context: https://caddy.community/t/path-regexp-vs-other-matchers/15186

We should have a more detailed reference for how it works and note the workarounds.

### PR DESCRIPTION
**Pull request by francislavoie**
Closes #214

Part of this assumes https://github.com/caddyserver/caddy/pull/5462 will be merged, so we might want to wait until that change is released before merging this.

### CODE REVIEW

File: src/docs/markdown/caddyfile/directives.md [hunk 1]
```diff
@@ -155,4 +159,33 @@ file_server
...
+    - `/foo` is more specific than `/foo*`
+    - `/foo/*` is more specific than `/foo*`
+
+  - A directive with any other matcher is sorted next, in the order it appears in the Caddyfile.
+
+    This includes path matchers with multiple values, and [named matchers](/docs/caddyfile/matchers#named-matchers).
+
+  - A directive with no matcher (i.e. matching all requests) is sorted last.
+
+- The [`vars`](/docs/caddyfile/directives/vars) directive has its ordering by matcher reversed, because it involves setting values which can overwrite eachother, so the most specific matcher should be evaluated last.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives.md:189]: I'd have to look but am sprinting through notifications atm -- is this also true of `map` and/or `header`? I can't recall off the top of my head so I just thought I'd make a quick note here.

**Review by mholt — APPROVED** [review]
This is an excellent improvement -- thank you! Just one question/nit I meant to check on but ran out of time.

File: src/docs/markdown/caddyfile/directives.md (hunk 1 above)
Comment (francislavoie) [reply src/docs/markdown/caddyfile/directives.md:189]: No we only put in that condition for `vars` currently. We could revisit for other directives though.

Comment (mholt) [reply src/docs/markdown/caddyfile/directives.md:189]: Ok; let's leave as-is for now. I guess I even use the `header` example in my wiki article, Composing in the Caddyfile: https://caddy.community/t/composing-in-the-caddyfile/8291

**Pull request merged**